RecoParticleFlow/PFProducer/interface/PFTables.h

```diff
@@ -0,0 +1,194 @@
+  PFClusterTables<edm::soa::ClusterTable, edm::soa::RecHitTable> clusters_hfhad_;
+  PFClusterTables<edm::soa::SuperClusterTable, edm::soa::SuperClusterRecHitTable> clusters_sc_;
+
+  // edm::soa::SuperClusterTable superclusters_table_;
+  // std::vector<size_t> element_to_supercluster_;
+
+  void clear() {
+    track_to_element_.clear();
+    element_to_track_.clear();
+    track_to_convrefs_.clear();
+
+    convref_table_.resize(0);
+    track_table_vertex_.resize(0);
+    track_table_ecalshowermax_.resize(0);
+    track_table_hcalent_.resize(0);
+    track_table_hcalex_.resize(0);
+    track_table_vfcalent_.resize(0);
+    track_table_ho_.resize(0);
+
+    gsf_table_.resize(0);
+    gsf_convbrem_table_.resize(0);
+    gsf_table_ecalshowermax_.resize(0);
+    gsf_table_hcalent_.resize(0);
+    gsf_table_hcalex_.resize(0);
+    gsf_to_element_.clear();
+    element_to_gsf_.clear();
+    gsf_to_convbrem_.resize(0);
+
+    element_to_brem_.clear();
+    brem_table_.resize(0);
+    brem_table_ecalshowermax_.resize(0);
+
+    clusters_ps1_.clear();
+    clusters_ps2_.clear();
+    clusters_hcal_.clear();
+    clusters_ecal_.clear();
+    clusters_ho_.clear();
+    clusters_hfem_.clear();
+    clusters_hfhad_.clear();
+    clusters_sc_.clear();
+
+    // superclusters_table_.resize(0);
+    // element_to_supercluster_.clear();
+  }
+
+  //this is needed since the KDTreeLinkers of different type need to have the same interface, but underneath may need to access different data
+  //according to KDTreeLinkerBase::_fieldType and KDTreeLinkerBase::_targetType
+  const PFClusterTables<edm::soa::ClusterTable, edm::soa::RecHitTable>& getClusterTable(
+      reco::PFBlockElement::Type type) const {
+    LogDebug("PFTables") << "getClusterTable " << type;
+    if (type == reco::PFBlockElement::HCAL) {
+      return clusters_hcal_;
+    } else if (type == reco::PFBlockElement::ECAL) {
+      return clusters_ecal_;
+    } else if (type == reco::PFBlockElement::HFEM) {
+      return clusters_hfem_;
+    } else if (type == reco::PFBlockElement::HFHAD) {
+      return clusters_hfhad_;
+    } else if (type == reco::PFBlockElement::HO) {
+      return clusters_ho_;
+    } else if (type == reco::PFBlockElement::PS1) {
+      return clusters_ps1_;
+    } else if (type == reco::PFBlockElement::PS2) {
+      return clusters_ps2_;
+    } else {
+      throw cms::Exception("unhandled type in getClusterTable");
+    }
+  }
+
+  const edm::soa::TrackTableExtrapolation& getTrackTable(reco::PFTrajectoryPoint::LayerType layer) const {
+    LogDebug("PFTables") << "getTrackTable " << layer;
+    if (layer == reco::PFTrajectoryPoint::HCALEntrance) {
+      return track_table_hcalent_;
+    } else if (layer == reco::PFTrajectoryPoint::HCALExit) {
+      return track_table_hcalex_;
+    } else if (layer == reco::PFTrajectoryPoint::VFcalEntrance) {
+      return track_table_vfcalent_;
+    } else {
+      throw cms::Exception("unhandler layer type in getTrackTable");
```

Comment:
```suggestion
      throw cms::Exception("unhandled layer type in getTrackTable");
```

Reply:
Note that this sets the exception _category_, not the _message_. Something along
```suggestion
      throw cms::Exception("LogicError") << "unhandled layer type in getTrackTable";
```
would provide clearer message (feel free to use better suited category).

Reply:
Thanks!